File: list/2/895D.py
def solve():
    import sys
    sys.setrecursionlimit(10**7)
    data = sys.stdin.read().split()
    if not data: 
        return
    a = data[0].strip()
    b = data[1].strip()
    n = len(a)
    mod = 10**9 + 7

    # Calcula a frequência de letras em a (o multiconjunto)
    freq_initial = [0] * 26
    for c in a:
        freq_initial[ord(c) - 97] += 1

    # Pré-calcula os fatoriais e invfatoriais até n
    fact = [1] * (n+1)
    invfact = [1] * (n+1)
    for i in range(1, n+1):
        fact[i] = fact[i-1] * i % mod
    invfact[n] = pow(fact[n], mod-2, mod)
    for i in range(n, 0, -1):
        invfact[i-1] = invfact[i] * i % mod

    # Retorna ways(rem, freq) = rem! / (Π (freq[l]!))
    def ways(rem, freq):
        res = fact[rem]
        for l in range(26):
            res = res * invfact[freq[l]] % mod
        return res

    # Conta quantas permutações (do multiconjunto) são lexicograficamente menores que x.
    def count_less(x, freq):
        total_rem = len(x)
        res = 0
        for i in range(len(x)):
            cur = ord(x[i]) - 97
            for l in range(cur):
                if freq[l] > 0:
                    freq[l] -= 1
                    res = (res + ways(total_rem - 1, freq)) % mod
                    freq[l] += 1
            if freq[cur] == 0:
                # x não pode ser formada (acabou a letra necessária)
                break
            freq[cur] -= 1
            total_rem -= 1
        return res

    # Calcula F(a) e F(b) usando cópias do vetor de frequências
    freq_a = freq_initial[:]  # cópia para calcular F(a)
    Fa = count_less(a, freq_a)
    freq_b = freq_initial[:]  # cópia para calcular F(b)
    Fb = count_less(b, freq_b)

    # O número de c tais que a < c < b é:
    # (número de permutações menores que b) - (número de permutações ≤ a)
    # Como a é uma permutação, seu rank (0-indexado) é F(a), assim:
    ans = (Fb - Fa - 1) % mod
    sys.stdout.write(str(ans))

if __name__ == '__main__':
    solve()

File: list/2/895D.cpp
#include <bits/stdc++.h>
using namespace std;
 
const long long MOD = 1000000007;
 
// Fast modular exponentiation.
long long modexp(long long base, long long exp) {
    long long result = 1;
    while(exp){
        if(exp & 1) result = (result * base) % MOD;
        base = (base * base) % MOD;
        exp >>= 1;
    }
    return result;
}
 
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
 
    string a, b;
    cin >> a >> b;
    int n = a.size();
    
    // Precompute factorials and inverse factorials up to n.
    vector<long long> fact(n+1), invfact(n+1);
    fact[0] = 1;
    for (int i=1; i<=n; i++){
        fact[i] = fact[i-1] * i % MOD;
    }
    invfact[n] = modexp(fact[n], MOD-2);
    for (int i = n; i >= 1; i--){
        invfact[i-1] = invfact[i] * i % MOD;
    }
 
    // count_less(x, freq): conta quantas permutações (do multiconjunto dado por freq) são < x.
    auto count_less = [&](const string &x, vector<int> freq) -> long long {
        long long res = 0;
        int rem = n;
        // cur_R = ∏_{l=0}^{25} invfact[freq[l]]
        long long cur_R = 1;
        for (int l = 0; l < 26; l++){
            cur_R = (cur_R * invfact[freq[l]]) % MOD;
        }
 
        for (int i = 0; i < n; i++){
            int cur = x[i] - 'a';
            int rem_minus = rem - 1;
            // Para cada letra menor que x[i] que esteja disponível...
            for (int l = 0; l < cur; l++){
                if(freq[l] > 0){
                    // Ao usar uma ocorrência da letra l, a nova contribuição do produto é:
                    // delta = invfact[freq[l]-1] * fact[freq[l]] mod MOD.
                    long long delta = (invfact[freq[l]-1] * fact[freq[l]]) % MOD;
                    res = (res + fact[rem_minus] * cur_R % MOD * delta) % MOD;
                }
            }
            if(freq[cur] == 0) break;
            int old_val = freq[cur];
            freq[cur]--;
            // Atualiza cur_R para refletir que usamos uma ocorrência de cur.
            cur_R = cur_R * (invfact[old_val-1] * fact[old_val] % MOD) % MOD;
            rem--;
        }
        return res;
    };
 
    vector<int> freq(26, 0);
    for (char c : a)
        freq[c - 'a']++;
 
    long long Fa = count_less(a, freq);
    long long Fb = count_less(b, freq);
    // Como a própria string a está no conjunto, seu “rank” (0-indexado) é Fa.
    // Queremos contar c tais que a < c < b, isto é, (Fb - Fa - 1).
    long long ans = (Fb - Fa - 1) % MOD;
    if(ans < 0) ans += MOD;
    cout << ans << "\n";
    return 0;
}
